conjure/ui/views/steps.py

```diff
@@ -1,60 +1,84 @@
 from conjure.ui.widgets.step import StepWidget
-from ubuntui.utils import Padding
+from ubuntui.utils import Padding, Color
 from ubuntui.widgets.table import Table
+from ubuntui.widgets.buttons import submit_btn, done_btn
 from urwid import WidgetWrap
 import random
 
 
 class StepsView(WidgetWrap):
 
-    def __init__(self, app, steps):
+    def __init__(self, app, steps, cb=None):
+        """ init
+
+        Arguments:
+        cb: process step callback
+        """
         self.app = app
         self.table = Table()
+        self.cb = cb
 
         self.steps_queue = steps
         for k, v in self.steps_queue.items():
-            self.steps_queue[k] = StepWidget(v)
+            meta = StepWidget(v['step_metadata'])
+            if not meta.viewable:
+                continue
             self.table.addColumns(
                 k,
                 [
-                    ('fixed', 3, self.steps_queue[k].icon),
-                    self.steps_queue[k].title,
+                    ('fixed', 3, meta.icon),
+                    meta.description,
                 ]
             )
-            self.table.addColumns(
-                k,
-                [
-                    ('weight', 1, Padding.left(self.steps_queue[k].result,
-                                               left=4))
-
-                ], force=True
-            )
+            for i in meta.additional_input:
+                self.table.addRow(Padding.line_break(""), False)
+                self.table.addColumns(
+                    k,
+                    [
+                        ('weight', 0.5, Padding.left(i['label'], left=5)),
+                        ('weight', 1, Color.string_input(
+                            i['input'],
+                            focus_map='string_input focus')),
+                    ], force=True
+                )
+                self.table.addRow(
+                    Padding.right_20(
+                        Color.button_primary(
+                            submit_btn(on_press=self.submit,
+                                       user_data=meta),
+                            focus_map='button_primary focus')), False)
+                self.table.addRow(Padding.line_break(""), False)
 
+        self.table.addRow(Padding.center_20(
+                        Color.button_primary(
+                            done_btn(on_press=self.done, label="View Summary"),
+                            focus_map='button_primary focus')))
         super().__init__(Padding.center_80(self.table.render()))
 
-    def update_step_message(self, key, msg):
-        """ Updates the return results from a step
-        """
-        self.steps_queue[key].result.set_text(('info_context', msg))
+    def done(self, *args):
+        self.cb({}, done=True)
+
+    def submit(self, btn, data):
+        self.cb(data)
```

Comment:
along the same lines as the comment above about 'meta', just calling this 'data' doesn't make it clear that it's a StepWidget object (at least I'm pretty sure it will be?)...

So when looking into what we're actually passing to finish() above, you need to look a lot deeper than if this had been `def submit(self, btn, stepwidget)`